### ISSUE TITLE
Deadlock when decoding TFRecords

### ISSUE THREAD

**Issue by hannes-brt** [tensorflow/tensorflow#6702]
I am storing my training examples as variable-length TFRecords using the following function for encoding:

```
def convert_to_tf_example(const_exonic_seq, const_intronic_seq,
                          alt_exonic_seq, alt_intronic_seq,
                          psi_distribution, psi_std,
                          alt_ss_position, alt_ss_type,
                          const_site_id, const_site_position,
                          n_alt_ss, event_type):
    """Encode a COSSMO example as a TFRecord"""

    assert len(alt_exonic_seq) == n_alt_ss
    assert len(alt_intronic_seq) == n_alt_ss
    # assert len(psi_distribution) == n_alt_ss
    # assert len(psi_std) == n_alt_ss
    assert event_type in ['acceptor', 'donor']
    assert len(alt_ss_type) == n_alt_ss
    assert all([t in ('annotated', 'gtex', 'maxent', 'hard_negative')
                for t in alt_ss_type])

    example = tf.train.SequenceExample(
        context=tf.train.Features(feature={
            'n_alt_ss': tf.train.Feature(
                int64_list=tf.train.Int64List(value=[n_alt_ss])
            ),
            'event_type': tf.train.Feature(
                bytes_list=tf.train.BytesList(value=[event_type])
            ),
            'const_seq': tf.train.Feature(
                bytes_list=tf.train.BytesList(
                    value=[const_exonic_seq, const_intronic_seq])
            ),
            'const_site_id': tf.train.Feature(
                bytes_list=tf.train.BytesList(
                    value=[const_site_id])
            ),
            'const_site_position': tf.train.Feature(
                int64_list=tf.train.Int64List(
                    value=[const_site_position])
            )
        }),
        feature_lists=tf.train.FeatureLists(feature_list={
            'alt_seq': tf.train.FeatureList(
                feature=[tf.train.Feature(
                    bytes_list=tf.train.BytesList(value=[aes, ais])
                ) for aes, ais in
                         zip(alt_exonic_seq, alt_intronic_seq)]
            ),
            'psi': tf.train.FeatureList(
                feature=[tf.train.Feature(
                    float_list=tf.train.FloatList(value=psi))
                         for psi in psi_distribution]),
            'psi_std': tf.train.FeatureList(
                feature=[tf.train.Feature(
                    float_list=tf.train.FloatList(value=psi_sd))
                         for psi_sd in psi_std]),
            'alt_ss_position': tf.train.FeatureList(
                feature=[tf.train.Feature(
                    int64_list=tf.train.Int64List(value=[pos]))
                         for pos in alt_ss_position]),
            'alt_ss_type': tf.train.FeatureList(
                feature=[tf.train.Feature(
                    bytes_list=tf.train.BytesList(value=[t]))
                        for t in alt_ss_type])
        })
    )
    return example
```

The data stored here is genomic, but the details shouldn't matter.

When I'm training, I use the following function to decode the TFRecords:
```
def read_single_cossmo_example(serialized_example, n_tissues):
    """Decode a single COSSMO example"""

    decoded_features = tf.parse_single_sequence_example(
        serialized_example,
        context_features={
            'n_alt_ss': tf.FixedLenFeature([], tf.int64),
            'event_type': tf.FixedLenFeature([], tf.string),
            'const_seq': tf.FixedLenFeature([2], tf.string),
            'const_site_id': tf.FixedLenFeature([], tf.string),
            'const_site_position': tf.FixedLenFeature([], tf.int64)
        },
        sequence_features={
            'alt_seq': tf.FixedLenSequenceFeature([2], tf.string),
            'psi': tf.FixedLenSequenceFeature([n_tissues], tf.float32),
            'psi_std': tf.FixedLenSequenceFeature([n_tissues], tf.float32),
            'alt_ss_position': tf.FixedLenSequenceFeature([], tf.int64),
            'alt_ss_type': tf.FixedLenSequenceFeature([], tf.string)
        }
    )
    return decoded_features
```

During training, I've been getting deadlocks with triple-digit CPU loads during training (running on a six-core i7) and I've isolated the problem to the above decoding function. A simple Tensorflow program like the following will reproduce the deadlock:

```
 with tf.device('/cpu:0'):
    filename_queue = tf.train.string_input_producer(
        train_files, num_epochs=num_epochs, shuffle=shuffle)
    file_reader = tf.TFRecordReader()
    tf_record_key, serialized_example = file_reader.read(filename_queue)
    decoded_example = read_single_cossmo_example(serialized_example,
                                                  n_tissues)

with tf.Session() as sess:
    sess.run(
        tf.variables_initializer(
            tf.global_variables() + tf.local_variables()
        )
    )

    # Start queue runners
    coord = tf.train.Coordinator()
    threads = tf.train.start_queue_runners(sess=sess, coord=coord)

    try:
        while not coord.should_stop():
            fetch_vals = sess.run(decoded_example)
    except tf.errors.OutOfRangeError:
        pass
    except KeyboardInterrupt:
        print "Training stopped by Ctrl+C."
    finally:
        coord.request_stop()
    coord.join(threads)
```

Now, setting `intra_op_threads = 1` and `inter_op_threads = 1` will prevent the small script from deadlocking. However, even when restricting the thread pool I have run into deadlocks when using these TFRecords in long running training sessions, so I suspect there is a deeper issue.

### Environment info
Operating System: Ubuntu 14.04

Installed version of CUDA and cuDNN: 
CUDA: 8.0
cuDNN: 5.1.5
If installed from binary pip package, provide:
https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-0.12.1-cp27-none-linux_x86_64.whl

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

**Comment by yaroslavvb** [comment]
Can you give a more complete reproducible example? I tried running your code on some synthetic data, and couldn't reproduce.

In particular, the function read_single_cossmo_example executes to completion. Do you mean that the problem happens on session.run on tensors created in that function?

Also, deadlock scenarios are typically low CPU usage -- you end up with every thread waiting on a mutex, and those don't take any CPU.

Some possibilities:

1. Parse example is inefficient and if you wait long enough, it'll produce the answer. One way to tease out inefficiencies is to use google profiling tools to get CPU profile, like is done in https://github.com/tensorflow/tensorflow/issues/6116

2. There's some bug in parse example which causes things to go into an infinite loop. In this case it would be useful to get the example which causes this condition. Also, perhaps you can gdb attach to the hanging process to see which function it is hanging in. If you compile with -c dbg, it'll give you line numbers as well, although it can make everything 10x slower and harder to reproduce

**Comment by hannes-brt** [comment]
Thanks @yaroslavvb,

the behaviour I'm seeing is kind of subtle, so hopefully I can describe it a little better.

The graph creation works fine, the "deadlock" happens in `session.run`, when running the graph. By deadlock I mean that CPU load gets extremely high, such that the machine becomes unresponsive.

This is quite certainly an issue with multi-threading. If I don't supply a ConfigProto to session.run, i.e. using the default parallelism settings, then I run into the deadlock after very iterations of the loop. 

If I set 
```
tf.Session(config=tf.ConfigProto(
        intra_op_parallelism_threads=1,
        inter_op_parallelism_threads=1,
        log_device_placement=True))
```
the script above should run forever without a problem.

_However,_ my problem is that even when I restricted parallelism, executing a long-running training script would still encounter the CPU load explosion situation which would crash my machine. I'm not sure how I could create a standalone reproducible example of this issue, so I'm trying to reduce the issue to the root cause and it seems to me that the fact that decoding the TFRecords fails with default parallelism settings is the original problem here.

I'm aware that `tf.parse_single_sequence_example` is inefficient, but there is no equivalent to `tf.parse_example` for sequence examples.

I will use the Google profiling tools to generate profiles in both the multi-threaded and single-threaded case. Hopefully, this will give some additional clues.

**Comment by yaroslavvb** [comment]
Could it be that there's a bad tfrecords example which causes infinite loop? If you use 1 thread, it takes longer to get to that example, so that explains the extra delay. You can fix seeds to help with determinism and print out example keys to see which example could be the problem

**Comment by hannes-brt** [comment]
I set a seed and printed every tf record key, but the it always hangs at a different example.

I did manage to run the CPU profiler though and created two profiles for the singled threaded configuration and the hanging multi threaded configuration.

Single-threaded: https://drive.google.com/open?id=0B_AiZCOgaorxSXRaaUl3SGJCYUk
Multi-threaded (hanging): https://drive.google.com/open?id=0B_AiZCOgaorxWlVxaUFnS1YzWWs

You can see that the hanging program spends all its time in `__sched_yield` in the Eigen thread pool.

**Comment by yaroslavvb** [comment]
Hm, from the profile it seems to get stuck in NonBlockingThreadPoolTempl::WorkerLoop -> sched_yield .... not really familiar with that part of the code but this vaguely seems like a bug in `parse_single_sequence_example` 

blame [says](https://github.com/tensorflow/tensorflow/blame/a4b9b3ad4d602621f2a81fe86d11a531aaa4701d/tensorflow/core/ops/parsing_ops.cc) vrv put that code in, but it maybe misattribution

**Comment by jart** [comment]
Per last comment, I'm assigning to @vrv to take this out of triage queue.

**Comment by vrv** [comment]
I don't really know what parse_single_sequence_example is, I think the original author was @ebrevdo, so assigning to him.

**Comment by ebrevdo** [comment]
That op does not use multiple threads.  Could you try running with
--config=asan to see if there are any associated memory issues?

If you want me to try to debug, you'll have to send me enough input data,
so i can reproduce, by email.  You can send it to my username@, the domain
is either gmail.com or google.com.

**Comment by hannes-brt** [comment]
@ebrevdo Emailed you with some scripts that should hopefully reproduce the problem

**Comment by hannes-brt** [comment]
@ebrevdo Have you been able to gain any insights?

**Comment by rajat95** [comment]
I am experiencing similar issue when trying to create a shuffled training batch. Here's my script for same. 

```

def batch_inputs(batch_size, train,sess, num_preprocess_threads=4,
                 num_readers=1):

    
  tf_record_pattern = os.path.join('/some_file',)
  data_files = tf.gfile.Glob(tf_record_pattern)
  if data_files is None:
    raise ValueError('No data files found for this dataset')
#        print data_files
        # Create filename_queue
  if train:
    filename_queue = tf.train.string_input_producer(data_files,
                                                          shuffle=True)
  else:
    filename_queue = tf.train.string_input_producer(data_files,
                                                          shuffle=False,
                                                          capacity=1)
  print filename_queue
  reader =tf.TFRecordReader()
  _, example_serialized = reader.read(filename_queue)

  image, label = parse_example_proto(example_serialized)

#  examples_per_shard = 201
  min_queue_examples = 20 * 2

  images, labels = tf.train.shuffle_batch(
    [image, label], batch_size=batch_size, num_threads=1,
    capacity=min_queue_examples + 3 * batch_size,
    min_after_dequeue=min_queue_examples)


  return images,labels


if __name__ == '__main__':

    sess = tf.Session()
    # imgs = tf.placeholder(tf.float32, [None,480, 800, 3])
    # labels = tf.placeholder(tf.float32, [None,2])
    # learning_rate = 0.005
    # vgg = vgg16(imgs, 'vgg16_weights.npz', sess)
    init_op = tf.group(tf.global_variables_initializer(),tf.local_variables_initializer())
    sess.run(init_op)
    coord = tf.train.Coordinator()
    threads = tf.train.start_queue_runners(sess=sess, coord=coord)
    images,labels  = batch_inputs(2,1,sess)
    i,l = sess.run([images,labels])
    print(i[0, :, :, :].shape)

    coord.request_stop()
    coord.join(threads)
```

**Comment by hannes-brt** [comment]
@ebrevdo @yaroslavvb After additional investigation I believe that this problem is OS-specific. I had a suspicion that the way the kernel queues tasks was related to the problem after I observed that the machine load would get extremely high (e.g. >1000 on a six core machine), but CPU, memory, and disk I/O would all remain low.
So I tried both my test script as well as my regular training task on Ubuntu 16.04 (from 14.04 before) and I haven't encountered the issue since.

**Comment by yaroslavvb** [comment]
nice investigation! Can you share more details for posterity? IE, what commands did you to use to query machine load (how is that different from CPU?)

**Comment by hannes-brt** [comment]
I mean load average as reported by e.g. `top` or `uptime` and many other tools. For example, here is an article that explains load and how it is different from CPU utilization: http://www.linuxjournal.com/article/9001. You can find people reporting similar problems by googling "high load average low cpu".

**Comment by girving** [comment]
Closing since this seems fixed by upgrading the OS.
